### DISCUSSION TITLE
Create subscriptionGroup through the command line and set the maximum number of retries for retryMaxTimes, which does not take effect：RocketMQ4.9.4

### DISCUSSION THREAD

**Discussion by onejimmyboy** [apache/rocketmq#6370] [Q&A]
create new group
./mqadmin updateSubGroup -b localhost:10911 -n localhost:9876 -r 1 -w 1 -s true -i 0 -c DefaultCluster -g test100
//consumer failed
DefaultMQPushConsumer consumer = new DefaultMQPushConsumer("group003",getAcl(),new AllocateMessageQueueAveragely());
consumer.setNamesrvAddr("localhost:9876");
consumer.subscribe("topic003","*");
consumer.registerMessageListener(new MessageListenerConcurrently() {
@OverRide
public ConsumeConcurrentlyStatus consumeMessage(List list,
ConsumeConcurrentlyContext consumeConcurrentlyContext) {
System.out.println("接收到的消息内容为：+"+ list);
return ConsumeConcurrentlyStatus.RECONSUME_LATER;
}
});
consumer.start();

What is expected to see?
Retry consumption 2 times

What did you see instead?
Retry consumption 16 times
Please tell us about your environment:
Rocketmq 4.9.4

After looking at the source code, it should be judged wrong here，
Although the maximum number of retries in the group is obtained here, it will be obtained from the header if the version is greater than 3.4.9. I don't know why this is done
![image](https://user-images.githubusercontent.com/57432151/225581528-2761ad2f-ba36-434d-a240-6b8b27cc9a2b.png)

**Comment by ShadowySpirits** [comment]
maxReconsumeTimesœ specified by the client takes precedence.

https://github.com/apache/rocketmq/blob/6fd1bb9a28ec1add82f7b85871b2c7c8876caa0b/client/src/main/java/org/apache/rocketmq/client/consumer/DefaultMQPushConsumer.java#L248-L255

**Reply by onejimmyboy** [reply]
用户在消费的时候，如果设置了这个maxReconsumeTimes ，我试了一下是没有问题的。看了一下上面的逻辑对于3.4.9以上的版本设置consumerGroup的最大重试次数不生效，以下的版本是生效的。但是在4.9.4版本，命令行是支持设置consumerGroup的最大重试次数的，我想知道在什么情况下这种对consumerGroup的设置才生效？

**Reply by ShadowySpirits** [reply]
The option `r` was added in UpdateSubGroupSubCommand since a very early version, but it is meanless now. IMO, we should remove this option.

<img width="1114" alt="image" src="https://user-images.githubusercontent.com/32436597/225793535-79f23d6e-4078-4f54-85ac-a028370281c7.png">

**Reply by onejimmyboy** [reply]
命令行删除这个可配置项也是一个解除误解的方法，反过来想一下，从consumerGroup的角度设置最大重试次数也是合理的，如果设置了，那每一个客户端就不用单独再设置了，作为同一设置的入口。如果客户端想单独设置，不使用consumerGroup的最大重试次数，也是可以自己设置的。不知道这种想法合理不？

**Reply by ShadowySpirits** [reply]
According to the java doc of `DefaultMQPushConsumer#maxReconsumeTimes`, maxReconsumeTimes has a default value of -1, which translates to 16 in concurrent mode. This strategy is implemented on the client side.

https://github.com/apache/rocketmq/blob/37ff93a27b034b2184144798d6c9f5bd00d63215/client/src/main/java/org/apache/rocketmq/client/impl/consumer/DefaultMQPushConsumerImpl.java#L867-L874

It is better to obey the subscription group config in the server when maxReconsumeTimes is not set by the user. However, given compatibility issues, we need more discussion.

**Reply by ShadowySpirits** [reply]
@aaron-ai What do you think? And is maxReconsumeTimes in the grpc client using the same way?

**Reply by drpmma** [reply]
The gRPC client uses the `retryMaxTimes` in `SubscriptionGroupConfig`.

See https://github.com/apache/rocketmq/blob/83c4a8f3f0284135d999c39a9fdc580b3d76d80e/proxy/src/main/java/org/apache/rocketmq/proxy/grpc/v2/common/GrpcClientSettingsManager.java#L102-L109

https://github.com/apache/rocketmq/blob/83c4a8f3f0284135d999c39a9fdc580b3d76d80e/proxy/src/main/java/org/apache/rocketmq/proxy/grpc/v2/common/GrpcClientSettingsManager.java#L140-L167
